### ISSUE TITLE
portability of sa data files

### ISSUE THREAD

**Issue by pschiffe** [sysstat/sysstat#135]
I'm not sure if this was discussed before and I know some of this already works, but in perfect reality, it would be possible to:

* read older sa data files with newer sysstat
    [this should be working, right? + there is a conversion with sadf -c]
* read newer sa data files with older sysstat
    [what's the current state? I know this is possible only to some extend, but would be great if this worked for future releases]
* sa data files from one architecture can be read by sysstat on another architecture (ppc64 vs x86_64 and so on)

I know that sa data files are binary, but... what do you think? What would be needed for that to work?

**Comment by sysstat** [comment]
This was never discussed before, but YES, this would be a great enhancement to sysstat. In fact, this is a feature that is absolutely necessary, even it means another (last?) break in the compatibility between sysstat versions.
As of today:
1) a newer sar can read older sa data files (to a certain extent, not too old though). For example, sar from 11.5.2 version can directly read sa data files from 10.3.1 and later versions. If you want to read an even older data file, you will need first to convert its format using sadf -c. With this option, you will be able to go back to 9.1.6. This is the oldest version you will ever be able to read (FYI, 9.1.6 was released in November 2010).
2) an older sar version can (under certain conditions) read newer sa data files: sar from 9.1.6 version can read sa data files up to 10.2.1, and sar from 10.3.1 version can read all the data files created by a newer sysstat version.
But when sar reads a newer data file containing an unknown activity (e.g., sar-10.3.1 doesn't know softnet stats added in 11.5.2) or a known activity with unknown new metrics (e.g., sar-10.3.1 doesn't know the field "kbavail" that is being added in next 11.5.3 version), it will skip the data without displaying them.
In fact, new activities won't bother an older version of sar. And as long as new metrics appear as new fields appended to the end of existing structures, this won't bother an older version of sar either.
Problems start when, for instance, the type of some kernel counters changes (e.g., going from "long int" to "long long int" on 32 bit architectures) which will impose a change of type in sysstat structures... and a break in compatibility format.
3) The last problem, as you mentioned it, is the portability between architectures of different endianness. Such a difference is detected by sar, but sar is unable to read such datafiles.

So what we would need is a self-describing data file format so that any versions of sar can read it and display its contents (which means that the name of the field, its type, whether it is an instantaneous value or a value calculated over the last interval of time, etc. should be recorded somewhere).
This file should also be independent of any kind of endianness, which means all the values are saved with a known endian type... or maybe even saved as strings of characters?
Last the file should be able to deal with a varying number of devices (network interfaces, disks, etc.) which can be registered dynamically by the system. So the file should accept any number of devices even if the number has grown and is much higher than the initial number when the file was created.

**Comment by pschiffe** [comment]
Yes, I was afraid that binary data files won't be good enough. One option is to use text based, self-described data files. In this case, I would try to use some existing format of structured data, like json or xml for better interoperability, maybe compressed by default.

Another option I'm thinking of is to use some simple database, like [sqlite](https://sqlite.org/). In this case, logic for the data storage could be delegated to the database engine, thus sysstat itself would be simpler in the end. This would also solve portability [between architectures](https://www.sqlite.org/appfileformat.html) and probably it would provide better performance than text based file format.
